contracts/epochs/FlowEpoch.cdc

```diff
@@ -406,6 +541,92 @@ pub contract FlowEpoch {
             FlowEpoch.account.load<Bool>(from: /storage/flowAutomaticRewardsEnabled)
             FlowEpoch.account.save(enabled, to: /storage/flowAutomaticRewardsEnabled)
         }
+        
+        /// Ends the currently active epoch and starts a new one with the provided configuration.
+        /// This function saves the recovery epoch metadata in storage containing the full specification for the new epoch.
+        /// This meta data will be emitted in the EpochRecover service event in the next heart beat interval.
+        /// This function is used within sporks to recover the network from Epoch Fallback Mode (EFM).
+        pub fun recoverEpoch(
+            randomSource: String,
+            startView: UInt64,
+            stakingEndView: UInt64,
+            endView: UInt64,
+            collectorClusters: [[String]],
+            clusterQCVoteData: [FlowClusterQC.ClusterQCVoteData],
+            dkgPubKeys: [String],
+            nodeIDs: [String])
+        {
+            pre {
+                FlowEpoch.isValidPhaseConfiguration(stakingEndView-startView+1, FlowEpoch.configurableMetadata.numViewsInDKGPhase, endView-startView+1):
+                    "Invalid startView, stakingEndView, and endView configuration"
+            }
+
+            if FlowEpoch.currentEpochPhase == EpochPhase.STAKINGAUCTION {
+                // Since we are resetting the epoch, we do not need to
+                // start epoch setup also. We only need to end the staking auction
+                FlowEpoch.borrowStakingAdmin().endStakingAuction()
+            } else {
+                // force reset the QC and DKG
+                FlowEpoch.borrowClusterQCAdmin().forceStopVoting()
+                FlowEpoch.borrowDKGAdmin().forceEndDKG()
+            }
+
+            // Construct the identity table for the recovery epoch
+            let nodes: [FlowIDTableStaking.NodeInfo] = []
+            
+            for nodeID in nodeIDs {
+                /// @TODO: if node is SN ensure weight > 0
+                /// throw exception, or add with 0 weight
+                nodes.append(FlowIDTableStaking.NodeInfo(nodeID))
+            }
+
+            // Compute the target end time for the next epoch
+            let timingConfig = FlowEpoch.getEpochTimingConfig()
+            let proposedTargetDuration = timingConfig.duration
+            let proposedTargetEndTime = timingConfig.getTargetEndTimeForEpoch(FlowEpoch.proposedEpochCounter())
+
+            // Create new Epoch metadata for the next epoch
+            // with the new values
+            let newEpochMetadata = EpochMetadata(
+                counter: FlowEpoch.currentEpochCounter + 1,
+                seed: randomSource,
+                startView: startView,
+                endView: endView,
+                stakingEndView: stakingEndView,
+                // This will be overwritten in `calculateAndSetRewards` below
```

Comment:
When you write: "This will be overwritten" I thought it only applies to the immediately next field. Though, I think you are changing multiple of the following field later, right? I that is the case, I would suggest:
```suggestion
                // The following fields will be overwritten in `calculateAndSetRewards` below
```